### PR TITLE
[DAT-126] - Add new authorization requirement to accept common users

### PR DESCRIPTION
**Pull request by bamaro-dev**
In this PR is added a new authorization requirement which allow access also to common user, only if user id matches with the user id that ask for the invoice.

- Im reviewing the unit test, because if the token expired, the new test fail. This is because if the token is expired, we cant see the token data.

Related to [DAT-126](https://makingsense.atlassian.net/browse/DAT-126)

Please, could you review it?

### CODE REVIEW

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, CanAccessUserRequirement requirement)
+        {
+            if (!IsSuperUser(context))
+            {
+                var tokenUserId = context.User.FindFirst(c => c.Type == ClaimTypes.NameIdentifier)?.Value;
+                var resource = context.Resource as Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext;
+                if (resource.RouteData.Values.TryGetValue("clientId", out var clientId) && clientId.ToString() != tokenUserId)
+                {
+                    _logger.LogWarning("The token Id and route Id are differents. The user hasn't permissions.");
```
Comment (jhoffman-ms) [inline Billing.API/DopplerSecurity/CanAccessUserHandler.cs:28]: Maybe the message should be: **"The IdUser into the token is different that in the route . The user hasn't permissions."** for me is more understandable. Do you think?

Comment (bamaro-dev) [reply Billing.API/DopplerSecurity/CanAccessUserHandler.cs:28]: Yes, maybe it's better to understand, you're right!

File: Billing.API.Test/InvoiceApiTest.cs [hunk 2]
```diff
@@ -48,10 +48,11 @@ public async Task GetInvoices_WhenInvalidPath_ReturnsNotFound(string url)
...
         [InlineData(HttpStatusCode.Unauthorized, "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9")]
+        [InlineData(HttpStatusCode.OK, "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1laWQiOjIyMjU0MSwidW5pcXVlX25hbWUiOiJiYW1hcm9AbWFraW5nc2Vuc2UuY29tIiwiaXNTdSI6ZmFsc2UsInN1YiI6ImJhbWFyb0BtYWtpbmdzZW5zZS5jb20iLCJjdXN0b21lcklkIjpudWxsLCJjZGhfY3VzdG9tZXJJZCI6bnVsbCwicm9sZSI6IlVTRVIiLCJpYXQiOjE2MDE5MTgyNzYsImV4cCI6MTYwMTkyMDA3Nn0.8J3SwjqB0GoTw3RwCeYwR9jAE_hZUnJ-1ooYYczl-JoBXBZYNxWjMpXl_JbYjC07rl0FW8lN1pas5ENWTA8WszDCTGSAu4YSFXcikleYOpddtuXBRGouM2PzRTWDm2ANY8yAScwVL0uzP3cG3u3819GYNWzoN1oC0rnDRvGmh2wXTid_mE-ZBq-DMuxnB1ivC1Ji728FfGtUkB0T8Qu9mYQnRCaNof8K47eWntko1kWG7LDxA83up3KxK-qhwnEwnNZQvhxneL8Q5SV5-qjtXfwn_8PBVGSqRZoc0tqeALfolzycCwlobmV9opYgbjqmgDQENBIyKV4kPfm8QS4JYg")]
```
Comment (nreal) [inline Billing.API.Test/InvoiceApiTest.cs:51]: This is a valid Token?

Comment (bamaro-dev) [reply Billing.API.Test/InvoiceApiTest.cs:51]: It was a valid token, but is expired now, so it's not valid anymore

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            if (!IsSuperUser(context))
+            {
+                var tokenUserId = context.User.FindFirst(c => c.Type == ClaimTypes.NameIdentifier)?.Value;
+                var resource = context.Resource as Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext;
+                if (resource.RouteData.Values.TryGetValue("clientId", out var clientId) && clientId.ToString() != tokenUserId)
+                {
+                    _logger.LogWarning("The IdUser into the token is different that in the route. The user hasn't permissions.");
+                    return Task.CompletedTask;
+                }
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/CanAccessUserHandler.cs:30]: Esto va a hacer que se deje pasar a los tokens personales a las URL que no tengan ClientId.

Que opinás de solo dejar pasar a los tokens personales cuando la URL tiene el parámetro ClientId y es igual al NameIdentifier del token? 

Por ahora creo que es lo mismo, pero en el futuro, si alguien hace un cambio, mejor tener por defecto lo más restrictivo.

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+            context.Succeed(requirement);
+            return Task.CompletedTask;
+        }
+
+        private bool IsSuperUser(AuthorizationHandlerContext context)
+        {
+            if (!context.User.HasClaim(c => c.Type.Equals("isSU")))
+            {
+                _logger.LogWarning("The token hasn't super user permissions.");
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/CanAccessUserHandler.cs:42]: Creo que ahora esto empieza a ser algo esperado, me parece que Warning no es el nivel de log correcto.

File: Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@ public static class DopplerSecurityServiceCollectionExtensions
     {
         public static IServiceCollection AddDopplerSecurity(this IServiceCollection services)
         {
-            services.AddSingleton<IAuthorizationHandler, IsSuperUserHandler>();
+            services.AddSingleton<IAuthorizationHandler, CanAccessUserHandler>();
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs:13]: You can keep both handlers ready to be used.

File: Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -19,7 +19,7 @@ public static IServiceCollection AddDopplerSecurity(this IServiceCollection serv
                 {
                     o.DefaultPolicy = new AuthorizationPolicyBuilder()
                         .AddAuthenticationSchemes(JwtBearerDefaults.AuthenticationScheme)
-                        .AddRequirements(new IsSuperUserRequirement())
+                        .AddRequirements(new CanAccessUserRequirement())
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs:22]: In my opinion, we can keep `IsSuperUserRequirement` as default requirement and add the `CanAccessUserRequirement` in a new policy only in the specific controller.

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace Billing.API.DopplerSecurity
+{
+    public class CanAccessUserHandler : AuthorizationHandler<CanAccessUserRequirement>
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/CanAccessUserHandler.cs:11]: IsSuperUserHandler

```suggestion
    public class IsSuperUserOrOwnResouceHandler : AuthorizationHandler<CanAccessUserRequirement>
```

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            _logger = logger;
+        }
+
+        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, CanAccessUserRequirement requirement)
+        {
+            if (!IsSuperUser(context))
+            {
+                var tokenUserId = context.User.FindFirst(c => c.Type == ClaimTypes.NameIdentifier)?.Value;
+                var resource = context.Resource as Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext;
+                if (resource.RouteData.Values.TryGetValue("clientId", out var clientId) && clientId.ToString() != tokenUserId)
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/CanAccessUserHandler.cs:26]: `resource` can be _null_, do not forget to validate it before trying to read `RouteData`.

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs (hunk 3 above)
Comment (bamaro-dev) [reply Billing.API/DopplerSecurity/CanAccessUserHandler.cs:30]: Buen punto! No lo tuve en cuenta porque el endpoint sin el `clientId `directamente no funciona, quiero decir, no encuentra el endpoint con una url sin `clientID`. Pero lo ajuste para tenerlo en cuenta, por si en algún momento se diera el caso, estar cubiertos.

File: Billing.API/DopplerSecurity/CanAccessUserHandler.cs (hunk 4 above)
Comment (bamaro-dev) [reply Billing.API/DopplerSecurity/CanAccessUserHandler.cs:42]: Es correcto. Lo cambie a LogDebug, de esa manera seguimos teniendo la info por si acaso tenemos que chequear si un token es super user.

File: Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs (hunk 6 above)
Comment (bamaro-dev) [reply Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs:22]: I tried that opcion, but if we keep IsSuperUserRequirement both requirements have to be met. The same happen if we add a new policy, both policies have to be met.
About "in the specific controller", maybe I lose something, but if I add that condicion in the controller, if IsSuperUserRequirement is not satisfied, never arrive to the endpoint. Have you another idea that I didnt see?

**Comment by andresmoschini** [comment]
🚀

**Pull request merged**